Add example of using @planetscale/database with Fastly Compute@Edge

JakeChampion opened the pull request:
This requires having the Fastly CLI installed (https://developer.fastly.com/learning/tools/cli/#installing) in order to build, run locally, and deploy to Fastly

mscoutermarsh commented on examples/fastly/README.md:
```diff
@@ -0,0 +1,16 @@
+# F1 Stats on Fastly Compute@Edge
+
+This is an example data source to retrive our dataset using Fastly Compute@Edge.
```
```suggestion
This is an example data source to retrieve our dataset using Fastly Compute@Edge.
```

mscoutermarsh reviewed (commented):
Thank you! we're going to try deploying & testing it out soon

mattrobenolt commented on examples/fastly/fastly.toml:
```diff
@@ -0,0 +1,23 @@
+service_id = ""
+
+[local_server.backends.psdb]
+url = "https://aws-eu-west-2.connect.psdb.cloud"
```
```suggestion
url = "https://aws.connect.psdb.cloud"
```

You'd want to use the global endpoint here.

mattrobenolt commented on examples/fastly/fastly.toml:
```diff
@@ -0,0 +1,23 @@
+  "PSCALE_PASSWORD" = ""
+
+[setup.backends.psdb]
+address = "aws-eu-west-2.connect.psdb.cloud"
```
```suggestion
address = "aws.connect.psdb.cloud"
```

Similarly here.

mattrobenolt commented on examples/fastly/src/index.js:
```diff
@@ -0,0 +1,79 @@
+ * @returns {Promise<Response>}
+ */
+async function app(event) {
+  const secrets = new Dictionary('secrets')
```
Just a question for my understanding, is this available on the global scope outside this function? If so, the better pattern would be to define:

```javascript
const db = new Client({ ... })
```

globally, then within the event handler, do `const conn = db.connection()`.

Concretely, this is what I was able to do for Vercel: https://github.com/planetscale/f1-championship-stats/blob/main/examples/vercel/api/data.json.ts#L9-L11

Some of the runtimes pass env vars and whatnot through the handler, so it's not really as feasible to have a global pool.

JakeChampion replied:
This is not constructable outside the handler as the top level scope is executed once during the wasm construction prior to uploading to fastly and so does not have access to the dictionary

mattrobenolt replied:
Gotcha, that makes sense!

mscoutermarsh commented on examples/fastly/README.md:
```diff
@@ -0,0 +1,16 @@
+## Setup
+
+1. Run `npm install`
+2. Run `npm run dev` to start the app locally.
```
Trying this out now. `npm run dev` isn't in the `package.json`, could you add that?

JakeChampion replied:
yes it should be 'fastly compute serve'

JakeChampion replied:
added in [`9eb3974` (#19)](https://github.com/planetscale/f1-championship-stats/pull/19/commits/9eb39748239a0992f091644c7b7a559c30852d01)

mscoutermarsh commented:
Got it deployed! It's currently 500'ing with this error.

```
stderr | f3f78115 | Error while running request handler: atob is not defined
stderr | f3f78115 | Stack:
stderr | f3f78115 |   decodeRow@<stdin>:190:20
stderr | f3f78115 |   parseRow@<stdin>:178:26
stderr | f3f78115 |   parse/<@<stdin>:187:38
stderr | f3f78115 |   parse@<stdin>:187:17
stderr | f3f78115 |   execute@<stdin>:123:29
stderr | f3f78115 |   async*app@<stdin>:259:14
stderr | f3f78115 |   @<stdin>:239:58
stderr | f3f78115 |
```

Also, would be helpful to add to the readme how to add the dictionary. I added one called `secrets` with the creds. This is my first time using fastly compute, so not sure if I did it correctly.

mattrobenolt commented:
I've opened an issue about base64 compatibility since this seems to be a hot one, and not only related to Fastly. https://github.com/planetscale/database-js/issues/49

@JakeChampion do you know if Fastly's runtime would have access to the `Buffer` API?

JakeChampion commented:
We don't have buffer but I am adding btoa and atob already 👍

JakeChampion commented:
> Also, would be helpful to add to the readme how to add the dictionary. I added one called `secrets` with the creds. This is my first time using fastly compute, so not sure if I did it correctly.

Apologies for that, I forgot to add the dictionary creation into `fastly.toml` - I will do that now and also address the other comments which have been raised 👍  Thank you for reviewing and deploying it to Fastly 🙌

JakeChampion commented:
> Also, would be helpful to add to the readme how to add the dictionary. I added one called `secrets` with the creds. This is my first time using fastly compute, so not sure if I did it correctly.

in commit [`9ed7c84` (#19)](https://github.com/planetscale/f1-championship-stats/pull/19/commits/9ed7c84f6e386e8c3b7743cc7607f7e69f701684) I added a section which will automatically create the dictionary with default values when the project is first deployed 👍

JakeChampion commented:
@mscoutermarsh and @mattrobenolt - this is ready for another round of reviews 🏄

mattrobenolt commented on examples/fastly/src/index.js:
```diff
@@ -0,0 +1,79 @@
+ * @returns {String}
+ */
+const ascii = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/='
+globalThis.btoa = function btoa(binary) {
```
@JakeChampion Did you end up still needing this? Or did this get added into the Fastly runtime? Because if it's still needed, you'd need to stub in `atob` as well, but I'm guessing if you've gotten this far, you don't need this.

JakeChampion replied:
It's not yet in the Fastly runtime, it should land in either the next release or the one after that 
I'll add a polyfill for `atob` as well and when Fastly's JS runtime supports both functions I can update the example to use the latest runtime and remove the polyfills 👍

JakeChampion replied:
This is the pull-request which implements btoa and atob in the Fastly runtime -> https://github.com/fastly/js-compute-runtime/pull/227

mattrobenolt reviewed (approved):
Aside from the `btoa` comment, this seems fine to me. I dunno if @mscoutermarsh had any last opinions.

mscoutermarsh commented:
I deployed it again to production and I'm getting

```
stderr | a536e7bf | Error while running request handler: atob is not defined
stderr | a536e7bf | Stack:
stderr | a536e7bf |   decodeRow@<stdin>:190:20
stderr | a536e7bf |   parseRow@<stdin>:178:26
stderr | a536e7bf |   parse/<@<stdin>:187:38
stderr | a536e7bf |   parse@<stdin>:187:17
stderr | a536e7bf |   execute@<stdin>:123:29
stderr | a536e7bf |   async*app@<stdin>:254:14
stderr | a536e7bf |   @<stdin>:239:58
stderr | a536e7bf |
```

on data.json

JakeChampion commented on examples/fastly/src/index.js:
```diff
@@ -0,0 +1,144 @@
+ * @param {String} data
+ * @returns {String}
+ */
+function aatob (data) {
```
```suggestion
globalThis.atob = function atob (data) {
```

JakeChampion commented:
> I deployed it again to production and I'm getting
> 
> ```
> stderr | a536e7bf | Error while running request handler: atob is not defined
> stderr | a536e7bf | Stack:
> stderr | a536e7bf |   decodeRow@<stdin>:190:20
> stderr | a536e7bf |   parseRow@<stdin>:178:26
> stderr | a536e7bf |   parse/<@<stdin>:187:38
> stderr | a536e7bf |   parse@<stdin>:187:17
> stderr | a536e7bf |   execute@<stdin>:123:29
> stderr | a536e7bf |   async*app@<stdin>:254:14
> stderr | a536e7bf |   @<stdin>:239:58
> stderr | a536e7bf |
> ```
> 
> on data.json

I had a silly typo where I named the function aatob instead of atob 🙈

mscoutermarsh commented:
🙌 getting closer. Deployed it again and now we're seeing

```
stderr | 9d3370fc | Error while running request handler: malformed UTF-8 character sequence at offset 1
stderr | 9d3370fc | Stack:
stderr | 9d3370fc |   decode@<stdin>:76:27
stderr | 9d3370fc |   cast@<stdin>:234:16
stderr | 9d3370fc |   parseRow/<@<stdin>:180:25
stderr | 9d3370fc |   parseRow@<stdin>:179:19
stderr | 9d3370fc |   parse/<@<stdin>:187:38
stderr | 9d3370fc |   parse@<stdin>:187:17
stderr | 9d3370fc |   execute@<stdin>:123:29
stderr | 9d3370fc |   async*app@<stdin>:258:14
stderr | 9d3370fc |   @<stdin>:239:58
stderr | 9d3370fc |
```

mattrobenolt commented:
Without trying to parse this, I'd suspect the polyfill is trying to decode base64 into utf-8 when it should be latin1 as per the definition of atob?

Both btoa and atob are explicitly latin1.

JakeChampion commented:
@mattrobenolt and @mscoutermarsh - This now uses the latest Fastly JS SDK which has native `atob` and `btoa` support 🎉

mscoutermarsh commented:
@JakeChampion I've got a strange bug. I deployed it and it was working for a couple minutes. But now whenever I hit it, I'm getting.

```
stderr | d1d7438a | Error while running request handler: btoa is not defined
stderr | d1d7438a | Stack:
stderr | d1d7438a |   postJSON@<stdin>:147:18
stderr | d1d7438a |   execute@<stdin>:117:27
stderr | d1d7438a |   app@<stdin>:254:14
stderr | d1d7438a |   @<stdin>:239:58
stderr | d1d7438a |
```

I've redeployed a few times and now getting that error 100% of the time. Any idea what's happening?

JakeChampion commented:
> @JakeChampion I've got a strange bug. I deployed it and it was working for a couple minutes. But now whenever I hit it, I'm getting.
> 
> ```
> stderr | d1d7438a | Error while running request handler: btoa is not defined
> stderr | d1d7438a | Stack:
> stderr | d1d7438a |   postJSON@<stdin>:147:18
> stderr | d1d7438a |   execute@<stdin>:117:27
> stderr | d1d7438a |   app@<stdin>:254:14
> stderr | d1d7438a |   @<stdin>:239:58
> stderr | d1d7438a |
> ```
> 
> I've redeployed a few times and now getting that error 100% of the time. Any idea what's happening?

It looks like the version is the JS runtime was not updated in the pull request - I went to rebase and update the version just now and GitHub got very confused and closed this pull request and refused to let me reopen it so I've made a new one which uses version 0.5.1 of the fastly js runtime (the version with atob and btoa)
